### DISCUSSION TITLE
How to use python to make a linux mobile system or phone? (By using micropython project)

### DISCUSSION THREAD

**Discussion by yingshaoxo** [micropython/micropython#12966] [General]
How to make a mobile phone and mobile phone system based on MIT license projects?

1. Prepare micro_python project. It is MIT licensed.
2. Change pyboard CPU to PC CPU. Change pyboard memory to 32GB, change pyboard storage to 2TB.
3. Make a never stop python loop. And let it run when connect to power. 
4. Use python to send signal and drive LED screen.
5. Use python to read keyboard or touch screen signal.
6. Use python to send sound signal to speakers.
7. Upgrade that python loop into a bash shell.
8. Let python be able to handle hardware USB signals.
9. Create pure image based graphic library with python, so it can sent out graphic signal stream directly to LED screen.
10. Use python to make a desktop launcher, which has terminal, file explorer, browser, 3 tabs.

#mobile #os #MIT #python #yingshaoxo

___

I tried to ask this question in https://unix.stackexchange.com, but it got deleted for unknown reason:

https://yingshaoxo.xyz/pictures/stackexchange/index.html

___

All I wish to see is someone make a useable mobile phone or system by using micropython, is that wrong?

**Comment by 2dof** [comment]
Question: You Want port micro-python project to linux/pc  mobile app ?

if yes, then for mobile apps written in Python (not micro python) You can use:  Kivy  or PyMob  (  Cross-platforms Python Framework))
 or  PyQt5 for GUI  + other frameworks/libs to create application

if You want use pure micro-python then should read about [[TTGO-T-watch-2020-Micropython-OS](https://github.com/jeffmer/TTGO-T-watch-2020-Micropython-OS)

**Comment by yingshaoxo** [comment]
@2dof 

I don't only want to create an apk or a mobile system for old devices

I want to create a MIT based ecosystem

I see the iPhone and Android system is getting more and more strict and unopened

For example, apple store and google play or new android system is trying to limit developers to use powerful permissions (For example, you can't publish another app store app in old store app)

And another reason is, android source code is around 60GB, almost about to explode. We need to cut it down to 50MB or less. (So anyone can do modification)

And another reason is the mobile hardware manufacturer is also not doing right thing, they encrypting everything about their hardware, for example you can't directly use camera, microphone, screen, NFC, GPS in your mobile device, you have to use manufacturer pre_compiled binary drivers. Which makes develop new system on old phone harder and harder. (If you have a 100MB mobile system, 80MB might be the drive binarys, it limits new system porting for different devices)

If we could create a new mobile phone hardware standards by using micropython project, everything would be easier.

**Comment by yingshaoxo** [comment]
Another thing I want to mention is:

If amd64 chip only needs 1.5 voltage to run, why those people use arm64 for mobile CPU?

I'm pretty sure everything will work OK with a amd64 chip.

**Reply by 2dof** [reply]
ARM is designed to run in low-power devices (mobile) - AMD - not.

**Comment by yingshaoxo** [comment]
## I have an idea of how to make a mobile desktop envirnoment for linux
First, we don't need those application in our platform to do anything but give us images when they think the user interface should get updated.

Then, what we need to do is send the keyboard, mouse, touchscreen, resolution to that application that is currently running at the front-end.

For each time, there would only have one single app is running at the front-end, I mean they can't send you image/graphic data at the background.

And for the desktop or launcher we have, there would always have 3 buttons at the bottom: recent_apps, home, back.

And what we will do as a desktop envirnoment is to send the mixed graphic data directly to the FTF SPI LCD Screen.

In this case, we don't need any modern desktop envirnoment that is normally greater than 1GB, but a simple python script that is smaller than 500KB.

### Protocol
#### For launcher
```
/add_or_update_app

request:
{
    "app_name": "",
    "app_icon_url": "",
    "app_url": ""
}
```

#### For each app
```
/get_graphic_image

request:
{
}

response:
{
    "has_image": true/false,
    "image": <base64_string>
}



/send_control_info

request:
{
    "keyboard": ascii_for_keyboard,
    "mouse": ...,
    "touch_screen": ...,,
    "resolution": 1080x1920,
}
```

**Reply by yingshaoxo** [reply]
I'm not making a desktop launcher, it is mobile only. And all design is resposive since I'll send resolution info to apps in realtime. (By the way, if user want to open two window in desktop, they could just do it, what is the difficulty? Render two window, one is left, another is right, they are all images.)

The multiple_window is not a problem, for those applications, they will always running in the backgroud, if user want to switch back to an app, they just have to use bottom 'recent app' to get recent app list, then click it to do a quick switch.

**Comment by yingshaoxo** [comment]
# About how to make the minimum web broswer:

According to my experience, the core html component is:

1. `<p></p>` for text
2. `<a></a>` for links
3. `<input></input>` for input, must have id
4. `<img></img>` for image, src could be base64 content
5. `<br>` for new line, `&nbsp;` for space character

When user input something in input tag, we save it in current page global dict, the key is input_id, then when user click any link at that page, we send that dict as url parameters along with that link.

When front-end browser show the content, it automatically do a scale based on the longest line.

JWT as an authentication method could always be in the `<a></a>` link so the user don't have to login again.

The whole thing is similar to wap web page.

#html #idea #yingshaoxo

**Reply by sophiedegran** [reply]
I think first you have to find a 4G LTE board which run well under micropython (with mic and small speaker)...
After making this challenge you'll see the consumption of all and if it's reliable to make an ecosystem around it...
I have seen some where a talky walky under Esp32 try to have a look at this project..

**Reply by yingshaoxo** [reply]
No, we don't need 4G board, we don't need SIM card related stuff. We need socket based realtime message,audio,video chat application. For example, livekit.

And by the way, I created a similar project two years ago, that's not a problem for me: https://github.com/weneedparty

**Comment by yingshaoxo** [comment]
# About how to make the minimal email system (message board system):

Create a web page, the upper 30% is the input box + send button, and the lower 70% is the message list.

Send a message through a GET request and obtain the message list. 

"http://api.com?message=***"

The server needs to remove duplicate messages.

#email #idea #yingshaoxo
